Transient thermal

JoranZwet opened the pull request:
Added two new modules:
- `TransientThermal`: Thermal solve for transient systems 
- `TransientToVTI`: Export transient responses to a vti.series

Added example to perform transient thermal compliance

Fixed few minor typos

JoranZwet committed:
added _prepare to Transient Thermal

JoranZwet committed:
added _prepare to Transient Thermal

JoranZwet committed:
added _response to Transient Thermal

JoranZwet committed:
added _response to Transient Thermal

JoranZwet committed:
added _sensitivity to Transient Thermal

JoranZwet committed:
added _sensitivity to Transient Thermal

JoranZwet committed:
added check for correct theta

JoranZwet committed:
added check for correct theta

JoranZwet committed:
small naming consistency change

JoranZwet committed:
small naming consistency change

JoranZwet committed:
bugfix in sensitivities

JoranZwet committed:
bugfix in sensitivities

JoranZwet committed:
added TransientThermal module

JoranZwet committed:
added TransientThermal module

JoranZwet committed:
small bugfixes

JoranZwet committed:
small bugfixes

JoranZwet committed:
fixed documentation

JoranZwet committed:
fixed documentation

JoranZwet committed:
Added example of transient thermal compliance

JoranZwet committed:
Added example of transient thermal compliance

JoranZwet committed:
added new modules

JoranZwet committed:
Added TransientToVTI module to allow exporting responses that vary over time

JoranZwet committed:
Added TransientToVTI module to allow exporting responses that vary over time

JoranZwet committed:
typo fix

JoranZwet committed:
typo fix

JoranZwet committed:
improved how solver is called

JoranZwet committed:
improved how solver is called

JoranZwet committed:
added documentation

JoranZwet committed:
added documentation

JoranZwet committed:
Small tweaks to documentation and file naming

JoranZwet committed:
Small tweaks to documentation and file naming

aatmdelissen commented on pymoto/modules/io.py:
```diff
@@ -1,9 +1,13 @@
+import sys
 from pathlib import Path
 import numpy as np
-import matplotlib
-matplotlib.use('TkAgg')  # Change default backend -- TkAgg does not freeze during calculations
+if platform.system() == 'Darwin':  # Avoid "Python is not installed as a framework (Mac OS X)" error
+    # Change backend
+    import matplotlib
+    matplotlib.use('TkAgg')
```
This seems to be changed wrongfully, maybe some issue with merging?

aatmdelissen commented on pymoto/modules/io.py:
```diff
@@ -352,3 +356,69 @@ def _response(self, *args):
+    cell-data or point-data in the VTI file. For 2D data (size is equal to ``2*nnodes``), the z-dimension is padded
+    with zeros to have 3-dimensional data. Also, non-transient block-vectors of multiple dimensions
+    (*e.g.* ``(2, 3*nnodes)``) are accepted, which get the suffixed as ``_00``.
+
```
I've tried to read the time data, but not sure how to use it in Paraview. I get `dat.xxxx.vti` files, like the other VTI writer module, and next to that there are folders with transient data. How do you use the data generated? Might be nice to add an explanation on that here.

aatmdelissen commented on pymoto/modules/io.py:
```diff
@@ -352,3 +356,69 @@ def _response(self, *args):
+
+    Args:
+        domain: The domain layout
+        transient_tags (list): Tag name of Signal(s) that vary over time
+        saveto (str): Location to save folders with transient responses for specific iterations
+        delta_t (float): Length of timestep
+        interval (int): Iteration interval for saving data
+        scale (float): Scaling factor for the domain
```
Typenames not required here, because they are already typehinted in the code. I tend to do put `(optional)` in case of optional arguments (`interval` and `scale`)

aatmdelissen commented on pymoto/modules/io.py:
```diff
@@ -352,3 +356,69 @@ def _response(self, *args):
+    def _response(self, *args):
+        if self.it % self.interval == 0:
+            #prepare folder of transient responses for specific optimization iteration
+            saveto = self.path[0] + '/transient_it{0:03d}/'.format(self.it) + self.path[1]
```
The filename contains the extension twice: `dat.vti.000.vti`. Maybe use `os.path.splitext` like in `WriteToVTI`.

aatmdelissen commented on pymoto/modules/io.py:
```diff
@@ -352,3 +356,69 @@ def _response(self, *args):
         self.iter += 1
+
+
+class TransientToVTI(Module):
```
Maybe name `SeriesToVTI` more general? Although I don't really know what other series than 'time' there would be...

aatmdelissen commented on pymoto/modules/io.py:
```diff
@@ -352,3 +356,69 @@ def _response(self, *args):
+
+            #prepare time series file
+            Path(saveto).parent.mkdir(parents=True, exist_ok=True)
+            f = open(saveto + '.series', 'a')
```
File is not closed when done writing. Use `with open(...) as f:` instead to have it automatically close.

aatmdelissen commented on pymoto/modules/transient.py:
```diff
@@ -0,0 +1,88 @@
+        solver: Manually override the LinearSolver used, instead of the solver from :func:`auto_determine_solver`
+        """
+    def _prepare(self, T_0, end, dt, theta = 1.0, solver = None):
+        self.T_0 = T_0
```
Can also be optional? (`zeros`)

aatmdelissen commented on pymoto/modules/transient.py:
```diff
@@ -0,0 +1,88 @@
+        """
+    def _prepare(self, T_0, end, dt, theta = 1.0, solver = None):
+        self.T_0 = T_0
+        self.steps = int(end/dt)
```
Can't this be deduced from the size of the `Q` matrix? Or maybe it would be nice to support a constant load-vector; in that case the number of steps would be required as user input.

JoranZwet replied:
I like supporting a constant load vector more in that case

aatmdelissen replied:
You could do both:
- If Q is constant, have the user provide some way to specify how many steps/end time
- If the Q is variable (so it has a second dimension with `N` steps) the number of steps will be overridden (the default value for the keyword argument would be `None`

aatmdelissen commented on pymoto/modules/transient.py:
```diff
@@ -0,0 +1,88 @@
+    def _sensitivity(self, dfdt):
+        # initialize adjoint variables
+        lams = np.zeros_like(self.temperature)
+        lams[:, -1] = self.solver.solve(-dfdt[:, -1])
```
No adjoint solve (transposed) required?

JoranZwet commented on pymoto/modules/io.py:
```diff
@@ -352,3 +356,69 @@ def _response(self, *args):
         self.iter += 1
+
+
+class TransientToVTI(Module):
```
Combining could definitely work. Would make the whole module a bit messy though as it would need a bunch of extra parameters and checks to see if it also needs to be saving certain optimization iterations to a transient folder

aatmdelissen replied:
I prefer messy if it makes the interface for the user simpler ;)

aatmdelissen commented on pymoto/modules/transient.py:
```diff
@@ -0,0 +1,88 @@
+    timestepping.
+
+    Input Signals:
+      - ``K`` (`dense or sparse matrix`): The system matrix :math:`\mathbf{K}` of size ``(n, n)``
+      - ``C`` (`dense or sparse matrix`): The damping matrix :math:`\mathbf{C}` of size ``(n, n)``
+      - ``Q`` (`dense matrix`): Right-hand-side matrix of size ``(n, Ntimesteps)``
```
We can keep the road open to future extension with mass matrix if we now already:
- [ ] Rename to something generic (e.g. `TransientSolve`)
- [ ] Put the load vector first, so the argument order will be `Q`, `K`, `D`, `M (optional)`

JoranZwet commented on pymoto/modules/io.py:
```diff
@@ -352,3 +356,69 @@ def _response(self, *args):
+
+            #loop over time steps, then generate new vtk file and add reference to series file for each time step
+            for t in range(data[self.transient_tags[0]].shape[1]):
+                for r in self.transient_tags:
```
possible, the one (edge-case) issue I think it would have is if you for whatever reason ran a transient simulation with only 3 timesteps you would run into problems with also recognizing vectors with 3 dofs as transient. Or as an extension of this same problem if you for some reason want to export X eigenvectors and also ran a transient simulation of X timesteps it would not distinguish between the two.

JoranZwet committed:
Changed:
- optional end time and initial state
- renamed variables to allow for later extension to second derivatives
- check for same matrix shape
- possibility to deduce total steps from input load

JoranZwet committed:
Changed name to TransientSolve

JoranZwet committed:
changed names to make it a general transient solve

JoranZwet committed:
changed names to make it use the general transient solve

JoranZwet committed:
fixed issues in SeriesToVTI

JoranZwet committed:
Adapted TransientSolve to new module structure

JoranZwet committed:
Adapted Transient example to new module structure

JoranZwet committed:
rebase on new root

JoranZwet committed:
bugfix in interval in WriteToVTI

JoranZwet committed:
small clarifying comment

JoranZwet committed:
Updated transient example from literature

JoranZwet committed:
Removed double lines from merging

JoranZwet committed:
Updated with new nomenclature

JoranZwet committed:
renamed to reflect new objective

JoranZwet committed:
Merge remote-tracking branch 'origin/transient' into transient

# Conflicts:
#	pymoto/__init__.py
#	pymoto/modules/io.py
#	pymoto/modules/linalg.py
#	pymoto/modules/transient.py

aatmdelissen reviewed (commented):
Looks good. Still some last small issues remaining.

JoranZwet committed:
removed old transient example

aatmdelissen committed:
Format and improve comments
